src/select/dropdown.js

```diff
@@ -16,171 +16,166 @@ import {StatefulMenu} from '../menu/index.js';
+  const highlightedIndex = React.useMemo(() => {
     // Highlight only first value as menu supports only a single highlight index
     let firstValue: OptionT = {};
 
-    if (Array.isArray(value) && value.length > 0) {
-      firstValue = value[0];
-    } else if (!(value instanceof Array)) {
-      firstValue = value;
+    if (Array.isArray(props.value) && props.value.length > 0) {
+      firstValue = props.value[0];
+    } else if (!(props.value instanceof Array)) {
+      firstValue = props.value;
     }
 
     if (Object.keys(firstValue).length > 0) {
-      const a = options.findIndex(
-        option => option && option[valueKey] === firstValue[valueKey],
+      const a = props.options.findIndex(
+        option =>
+          option && option[props.valueKey] === firstValue[props.valueKey],
       );
       return a || 0;
     }
     return 0;
-  };
+  }, [props.value, props.options, props.valueKey]);
 
-  render() {
-    // TODO(#185) Add no-results and loading states to menu
-    const {
-      maxDropdownHeight,
-      multi,
-      noResultsMsg,
-      onItemSelect,
-      options = [],
-      overrides = {},
-      size,
-    } = this.props;
-    const [DropdownContainer, dropdownContainerProps] = getOverrides(
-      overrides.DropdownContainer,
-      StyledDropdownContainer,
-    );
-    const [ListItem, listItemProps] = getOverrides(
-      overrides.DropdownListItem,
-      StyledDropdownListItem,
+  const groupedOptions = React.useMemo(() => {
+    return props.options.reduce(
+      (groups, option) => {
+        if (option.__optgroup) {
+          if (!groups[option.__optgroup]) {
+            groups[option.__optgroup] = [];
+          }
+          groups[option.__optgroup].push(option);
+        } else {
+          groups.__ungrouped.push(option);
+        }
+        return groups;
+      },
+      {__ungrouped: []},
     );
-    const [
-      OverriddenStatefulMenu,
-      // $FlowFixMe
-      {overrides: statefulMenuOverrides = {}, ...restStatefulMenuProps},
-    ] = getOverrides(overrides.StatefulMenu, StatefulMenu);
-    const highlightedIndex = this.getHighlightedIndex();
-    return (
-      <DropdownContainer
-        ref={this.props.innerRef}
-        role="listbox"
-        {...this.getSharedProps()}
-        {...dropdownContainerProps}
-      >
-        <OverriddenStatefulMenu
-          noResultsMsg={noResultsMsg}
-          onItemSelect={onItemSelect}
-          items={options}
-          size={size}
-          initialState={{
-            isFocused: true,
-            highlightedIndex: highlightedIndex,
-          }}
-          overrides={mergeOverrides(
-            {
-              List: {
-                component: StyledDropdown,
-                style: p => ({
-                  maxHeight: p.$maxHeight || null,
-                }),
-                props: {
-                  $maxHeight: maxDropdownHeight,
-                  'aria-multiselectable': multi,
-                },
+  }, [props.options]);
```

Comment:
this makes assumption that `props.options` is immutable but usually it's nots?

Reply:
yeah good point, i think i was getting a bit too aggressive on this perf optimization without truly measuring the improvements. i'll run some tests tommorow